modules/backup-azure/client.go

```diff
@@ -0,0 +1,208 @@
+)
+
+type azureClient struct {
+	client     *azblob.Client
+	config     clientConfig
+	serviceURL string
+	dataPath   string
+}
+
+func newClient(ctx context.Context, config *clientConfig, dataPath string) (*azureClient, error) {
+	connectionString := os.Getenv("AZURE_STORAGE_CONNECTION_STRING")
+	if connectionString != "" {
+		client, err := azblob.NewClientFromConnectionString(connectionString, nil)
+		if err != nil {
+			return nil, errors.Wrap(err, "create client using connection string")
+		}
+		serviceURL := ""
+		connectionStrings := strings.Split(connectionString, ";")
+		for _, str := range connectionStrings {
+			if strings.HasPrefix(str, "BlobEndpoint") {
+				blobEndpoint := strings.Split(str, "=")
+				if len(blobEndpoint) > 1 {
+					serviceURL = blobEndpoint[1]
+					if !strings.HasSuffix(serviceURL, "/") {
+						serviceURL = serviceURL + "/"
+					}
+				}
+			}
+		}
+		return &azureClient{client, *config, serviceURL, dataPath}, nil
+	}
+
+	// Your account name and key can be obtained from the Azure Portal.
+	accountName := os.Getenv("AZURE_STORAGE_ACCOUNT_NAME")
+	accountKey := os.Getenv("AZURE_STORAGE_PRIMARY_ACCOUNT_KEY")
+
+	if accountName == "" {
+		return nil, errors.New("AZURE_STORAGE_ACCOUNT_NAME must be set")
+	}
+
+	// The service URL for blob endpoints is usually in the form: http(s)://<account>.blob.core.windows.net/
+	serviceURL := fmt.Sprintf("https://%s.blob.core.windows.net/", accountName)
+
+	if accountKey != "" {
+		cred, err := azblob.NewSharedKeyCredential(accountName, accountKey)
+		if err != nil {
+			return nil, err
+		}
+
+		client, err := azblob.NewClientWithSharedKeyCredential(serviceURL, cred, nil)
+		if err != nil {
+			return nil, err
+		}
+		return &azureClient{client, *config, serviceURL, dataPath}, nil
+	}
+
+	client, err := azblob.NewClientWithNoCredential(serviceURL, nil)
+	if err != nil {
+		return nil, err
+	}
+	return &azureClient{client, *config, serviceURL, dataPath}, nil
+}
+
+func (a *azureClient) HomeDir(backupID string) string {
+	return a.serviceURL + path.Join(a.config.Container, a.makeObjectName(backupID))
+}
+
+func (a *azureClient) makeObjectName(parts ...string) string {
+	base := path.Join(parts...)
+	return path.Join(a.config.BackupPath, base)
+}
+
+func (a *azureClient) GetObject(ctx context.Context, backupID, key string) ([]byte, error) {
+	objectName := a.makeObjectName(backupID, key)
+
+	blobDownloadResponse, err := a.client.DownloadStream(ctx, a.config.Container, objectName, nil)
+	if err != nil {
+		if bloberror.HasCode(err, bloberror.BlobNotFound) {
+			return nil, backup.NewErrNotFound(errors.Wrapf(err, "get object '%s'", objectName))
+		}
+		return nil, backup.NewErrInternal(errors.Wrapf(err, "download stream for object '%s'", objectName))
+	}
+
+	reader := blobDownloadResponse.Body
+	downloadData, err := io.ReadAll(reader)
+	if err != nil {
+		return nil, backup.NewErrInternal(errors.Wrapf(err, "read stream for object '%s'", objectName))
```

Comment:
In case of error we need to Make sure` reader.Close()`

Reply:
I have addressed this issue and I'm closing the reader right away. Thank you for pointing this.